### PR TITLE
[WIP] Add some new pytest features

### PR DESCRIPTION
**Pull request by mitzkia**
This PR allows to do the followings in pytest fw:

syslog-ng config related
- Support log { flags(...)} from config
- Support filters {...} from config
- Support embedded log statements in config

syslog-ng cli related
- if syslog-ng crashed and core file was generated detect and save core files with backtrace
- Allow to stop syslog-ng with signal
- Support to run syslog-ng behind external tools: strace, perf, valgrind

some bugfixes
- bugfix: Save message_reader between readings
- bugfix: Fix running tests with marked 'slow'

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: tests/pytest_framework/functional_tests/profiling/test_with_valgrind.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def process_valgrind_log(valgrind_log_path):
+    with open(valgrind_log_path, "r") as valgrind_log:
+        valgrind_content = valgrind_log.read()
+        assert "Invalid read" not in valgrind_content
+        assert "Invalid write" not in valgrind_content
+
+
+@pytest.mark.slow
```
Comment (furiel) [inline tests/pytest_framework/functional_tests/profiling/test_with_valgrind.py]: That would be so awesome if we had a `@valgrind` decorator instead, not sure how difficult to write such a decorator though (or even possible?).

File: tests/pytest_framework/functional_tests/profiling/test_with_valgrind.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    for __counter in range(1, 5):
+        syslog_ng.reload(config)
+        file_source.write_log(bsd_log, counter=100)
+        syslog_ng.reload(config)
+
+    syslog_ng.stop()
+
+    valgrind_log_path = syslog_ng.get_instance_paths().get_valgrind_log_path()
+    process_valgrind_log(valgrind_log_path)
```
Comment (furiel) [inline tests/pytest_framework/functional_tests/profiling/test_with_valgrind.py]: Similar to core dumps, this should be checked inside stop, if possible.

File: tests/pytest_framework/src/syslog_ng/syslog_ng.py [hunk 3]
```diff
@@ -27,22 +27,26 @@
...
 class SyslogNg(object):
     def __init__(self, logger_factory, instance_paths, syslog_ng_ctl):
+        self.instance_paths = instance_paths
         self.__syslog_ng_cli = SyslogNgCli(
             logger_factory, instance_paths, ConsoleLogReader(logger_factory, instance_paths), syslog_ng_ctl
         )
 
-    def start(self, config):
-        self.__syslog_ng_cli.start(config)
+    def start(self, config, external_tool=None):
```
Comment (furiel) [inline tests/pytest_framework/src/syslog_ng/syslog_ng.py]: Back to the decorator-ish idea: what if the external tool were stored inside tc? Probably we could decorate tc to store it. Then this variant would be available for the constructors, and passed silently through the call chain without changing the user api at all (modulo the `@valgrind` decorator of course).

File: tests/pytest_framework/src/syslog_ng/syslog_ng_paths.py [hunk 4]
```diff
@@ -50,6 +50,9 @@ def set_syslog_ng_paths(self, instance_name):
                 "control_socket_path": Path(relative_working_dir, "syslog_ng_{}.ctl".format(instance_name)),
                 "stderr": Path(working_dir, "syslog_ng_{}_stderr".format(instance_name)),
                 "stdout": Path(working_dir, "syslog_ng_{}_stdout".format(instance_name)),
+                "valgrind": Path(working_dir, "valgrind.log"),
```
Comment (furiel) [inline tests/pytest_framework/src/syslog_ng/syslog_ng_paths.py]: Shouldn't we randomize path here too?

File: tests/pytest_framework/src/syslog_ng/syslog_ng_paths.py [hunk 5]
```diff
@@ -96,3 +99,12 @@ def get_syslog_ng_bin(self):
 
     def get_syslog_ng_ctl_bin(self):
         return self.__syslog_ng_paths["binary_file_paths"]["syslog_ng_ctl"]
+
+    def get_strace_log_path(self):
```
Comment (furiel) [inline tests/pytest_framework/src/syslog_ng/syslog_ng_paths.py]: Unused functions.

File: tests/pytest_framework/src/syslog_ng/tests/test_syslog_ng_paths.py [hunk 6]
```diff
@@ -38,6 +38,9 @@ def test_syslog_ng_paths(tc_unittest):
         "control_socket_path",
         "stderr",
         "stdout",
+        "valgrind",
```
Comment (furiel) [inline tests/pytest_framework/src/syslog_ng/tests/test_syslog_ng_paths.py]: Do we need these list elements?

File: tests/pytest_framework/src/syslog_ng_config/renderer.py [hunk 7]
```diff
@@ -41,6 +41,8 @@ def __render(self, re_create_config=None):
             self.__render_version()
         if self.__syslog_ng_config["sources"]:
             self.__render_statements(root_statement="sources", statement_name="source")
+        if self.__syslog_ng_config["filters"]:
```
Comment (furiel) [inline tests/pytest_framework/src/syslog_ng_config/renderer.py]: For now this looks good, but I am afraid we will run into trouble when rewrite rules are introduced, because they freely in mix into filters (ordering is undefined).

File: tests/pytest_framework/src/syslog_ng_config/statements/filters/filter.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+#############################################################################
+
+from src.common.random_id import RandomId
+
+
+class Filter:
+    def __init__(self, logger_factory, instance_paths, **kwargs):
+        self.options = kwargs
+
+        self.driver_id = "driverid_%s" % RandomId(use_static_seed=False).get_unique_id()
```
Comment (furiel) [inline tests/pytest_framework/src/syslog_ng_config/statements/filters/filter.py]: Instead of `"driverid_%s"`, how about `"filterid_%s"`?

Comment (furiel) [inline tests/pytest_framework/src/syslog_ng_config/statements/filters/filter.py]: Why do we need to disable the static seed?

File: tests/pytest_framework/src/syslog_ng_config/statements/logpath/logpath.py [hunk 9]
```diff
@@ -27,13 +27,17 @@
...
         self.logpath_id = "logpath_%s" % RandomId(use_static_seed=False).get_unique_id()
-        self.full_logpath_node = {self.logpath_id: {"sources": [], "destinations": [], "flags": []}}
+        self.full_logpath_node = {self.logpath_id: {"sources": [], "filters": [], "destinations": [], "flags": []}}
         self.logpath_node = self.full_logpath_node[self.logpath_id]
 
     def add_source_groups(self, source_groups):
         target_node = self.logpath_node["sources"]
         self.update_logpath_node(target_node, source_groups)
 
+    def add_filter_groups(self, filter_groups):
```
Comment (furiel) [inline tests/pytest_framework/src/syslog_ng_config/statements/logpath/logpath.py]: A pattern is evolving here too. These four functions do identical things except a string that specifies the target node. Please eliminate the repetition.

File: tests/pytest_framework/src/syslog_ng/syslog_ng.py [hunk 10]
```diff
@@ -35,8 +35,8 @@ def __init__(self, logger_factory, instance_paths, syslog_ng_ctl):
     def start(self, config, external_tool=None):
         self.__syslog_ng_cli.start(config, external_tool)
 
-    def stop(self, unexpected_messages=None):
-        self.__syslog_ng_cli.stop(unexpected_messages)
+    def stop(self, unexpected_messages=None, signal=None):
```
Comment (furiel) [inline tests/pytest_framework/src/syslog_ng/syslog_ng.py]: This patch is not clear for me. I have a few questions for clarifications, because I do not see the use case here.
- Why do we want to stop syslog-ng with signal.
- How can the users utilize this? Can you add an example testcase with assertions.
- Do we want to allow any kind of signals, like sigchild?
- Can we avoid extending the `stop` method with this parameter?

It would worth put this patch into a separate PR.

File: tests/pytest_framework/src/syslog_ng_config/syslog_ng_config.py [hunk 11]
```diff
@@ -81,15 +81,31 @@ def create_filter_group(self, filters):
...
         if destinations:
             logpath.add_destination_groups(destinations)
+        if logpaths:
+            logpath.add_logpaths(logpaths)
         if flags:
             logpath.add_flags(flags)
         self.__syslog_ng_config["logpaths"].update(logpath.full_logpath_node)
         return logpath
+
+    def create_inner_logpath(self, sources=None, filters=None, destinations=None, logpaths=None, flags=None):
```
Comment (furiel) [inline tests/pytest_framework/src/syslog_ng_config/syslog_ng_config.py]: Can you separate this patch too in a separate PR? I miss an example here as well.

File: tests/pytest_framework/src/syslog_ng_config/statements/filters/filter.py (hunk 8 above)
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng_config/statements/filters/filter.py]: +1, will be fixed.

Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng_config/statements/filters/filter.py]: With enabled static seed in every case random ids would be the same, imagine:
source sourceid_27096 { file(); };
source sourceid_27096 { file(); };

File: tests/pytest_framework/src/syslog_ng_config/statements/logpath/logpath.py (hunk 9 above)
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng_config/statements/logpath/logpath.py]: +1, will be fixed.

File: tests/pytest_framework/src/syslog_ng_config/syslog_ng_config.py (hunk 11 above)
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng_config/syslog_ng_config.py]: +1, separate PR is coming

File: tests/pytest_framework/src/syslog_ng/syslog_ng_cli.py [hunk 12]
```diff
@@ -91,7 +93,7 @@ def reload(self, config):
 
     def stop(self, unexpected_messages=None):
         self.__logger.info("Beginning of syslog-ng stop")
-        if self.__process:
+        if self.__process and not self.__core_file_exist:
```
[the remark this answers is not in the document]
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng/syslog_ng_cli.py]: Ok, I will check this case. thx

File: tests/pytest_framework/src/syslog_ng/syslog_ng_cli.py [hunk 13]
```diff
@@ -112,3 +114,12 @@ def stop(self, unexpected_messages=None):
     # Helper functions
     def __error_handling(self):
         self.__console_log_reader.dump_stderr()
+        self.__check_core_file(self.__process)
+
+    def __check_core_file(self, process):
```
[the remark this answers is not in the document]
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng/syslog_ng_cli.py]: +1, will be fixed.

File: tests/pytest_framework/src/syslog_ng/syslog_ng_cli.py [hunk 14]
```diff
@@ -112,3 +114,12 @@ def stop(self, unexpected_messages=None):
...
+        if process.wait(1) != 0:
+            for core_file in Path(".").glob("*core*"):
+                self.__core_file_exist = True
+                self.__syslog_ng_executor.get_backtrace_from_core(core_file=str(core_file))
```
[the remark this answers is not in the document]
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng/syslog_ng_cli.py]: - "." means the dir from where you started testcase (pytest), usually in my case: syslog-ng source/tests/pytest-framework/
- get_working_dir() correct, this means the test log directory

File: tests/pytest_framework/src/syslog_ng/syslog_ng_cli.py [hunk 15]
```diff
@@ -112,3 +114,12 @@ def stop(self, unexpected_messages=None):
...
+                core_file.replace(Path(self.__instance_paths.get_working_dir(), core_file))
+                raise Exception("syslog-ng core file found and processed")
```
[the remark this answers is not in the document]
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng/syslog_ng_cli.py]: +1, will be fixed

File: tests/pytest_framework/src/syslog_ng/syslog_ng_executor.py [hunk 16]
```diff
@@ -45,6 +45,22 @@ def run_command(self, command_short_name, command):
...
+            "-ex",
+            "bt full",
+            "--batch",
+            self.__instance_paths.get_syslog_ng_bin(),
+            "--core",
+            core_file,
+        ]
+        return self.__command_executor.run(
+            command=gdb_command_args,
+            stdout_path=self.__instance_paths.get_stdout_path_with_postfix(postfix="gdb_core"),
```
[the remark this answers is not in the document]
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng/syslog_ng_executor.py]: +1, will be fixed

File: tests/pytest_framework/src/syslog_ng/syslog_ng.py (hunk 10 above)
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng/syslog_ng.py]: I am also not really sure to support this case.
Example testcase:
import signal
config = some_dummy_config()
syslog_ng = tc.new_syslog_ng()
syslog_ng.start(config)
syslog_ng.stop(signal=signal.SIGSEGV)

With such steps we can test if syslog-ng saves internal states/persist correctly. 

Is it ok?

Comment (furiel) [reply tests/pytest_framework/src/syslog_ng/syslog_ng.py]: For that case, I think creating a `syslog_ng.send_signal(signal.SIGSEGV)` would be more clear. Syslog-ng does not necessarily stop for any kind of signals, so linking signal management with stop is not a good coupling.

File: tests/pytest_framework/src/syslog_ng_config/renderer.py [hunk 17]
```diff
@@ -89,4 +89,6 @@ def __render_logpath(self):
                 self.__syslog_ng_config_content += "    source({});\n".format(src_driver)
             for dst_driver in self.__syslog_ng_config["logpaths"][logpath]["destinations"]:
                 self.__syslog_ng_config_content += "    destination({});\n".format(dst_driver)
+            for flags in self.__syslog_ng_config["logpaths"][logpath]["flags"]:
```
[the remark this answers is not in the document]
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng_config/renderer.py]: +1

File: tests/pytest_framework/src/syslog_ng/syslog_ng.py (hunk 3 above)
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng/syslog_ng.py]: +1, like the idea

File: tests/pytest_framework/src/syslog_ng/syslog_ng_paths.py (hunk 4 above)
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng/syslog_ng_paths.py]: +1, will be fixed

File: tests/pytest_framework/src/syslog_ng_config/renderer.py (hunk 7 above)
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng_config/renderer.py]: Correct, we should think about this later.

File: tests/pytest_framework/src/syslog_ng/syslog_ng.py (hunk 10 above)
Comment (mitzkia) [reply tests/pytest_framework/src/syslog_ng/syslog_ng.py]: +1, will be fixed for this way

**Comment by mitzkia** [comment]
Thanks to Antal for the lot of questions and review notes.
Some of them I have already answered and some of them I have fixed.
Because this PR is much bigger than I planned in the beggining
(Antal also mentioned it), I decided to split it into 4 new PRs, and I am
closing this PR.
New PRs:
#2440 
#2441
#2442 
#2443
